[ITZXITZX] iP

ITZXITZX opened the pull request:
# BobPro

> "Your mind is for having ideas, not holding them." – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

BobPro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,
1. download it from it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ITZXITZX committed:
Renamed Duke to Bob and changed print output.

ITZXITZX committed:
Updated Bob class to echo user input.

Bob class could only print a standard output.

Add 1 main chat function and 1 print helper function.

Extracting common print output format reduces repeated manual code that
may give inconsistent output. Chat is now responsive to user input.

Let's,
* include a chat function that echos user input.
* include a helper print function that formats the printed output.

ITZXITZX committed:
listRecords function: list a record of all user inputs

Able to return a list of all user inputs

Let's
* add counter to store the latest count of input
* add records to store all records in an Arraylist
* add getters and setters to retrieve and update these fields
* add listRecord function to list all records
* update the chat function to store each input by the user

ITZXITZX committed:
Marking whether a task is completed.

Previously unable to track task completion.

Defined Task class to handle each task.

ITZXITZX committed:
Trimmed new inputs in the while loop.

ITZXITZX committed:
Bob class is using getters to access Task class.

There is a lack of information hiding.

Removal of getter methods, and solely having 1 mark task method inside
the Task class.

This is to allow for information hiding.

ITZXITZX committed:
There is a lot of repeated code in the chat function.

Functions where identified and refactored from the main chat function
logic.

Better separation of tasks.

ITZXITZX committed:
Unable to keep track of different categories of tasks.

To record important information for types of tasks.

Subclasses of Tasks created.

Inheritance allows for better categorisation of tasks.

ITZXITZX committed:
Manual Regression Testing.

Takes too much time to test for any regressions.

Automated Testing of text UI.

Compares the actual and expected results faster.

ITZXITZX committed:
Little Error Handling.

Checked Errors from user inputs need to be handled by the program.

InvalidTaskException created and used to handle invalid inputs.

Handles requirement specific errors.

ITZXITZX committed:
Unable to delete tasks.

Add delete method to Bob class.

Allows deletion of task from the records of task within Bob.

ITZXITZX committed:
Added description to isValidTaskNumber.

ITZXITZX committed:
Another edit to the description of isValidTaskNumber.

ITZXITZX committed:
Added description to delete method.

ITZXITZX committed:
Merge branch 'master' into test1

# Conflicts:
#	src/main/java/Bob.java

ITZXITZX committed:
Removed commented out line.

ITZXITZX committed:
Removed empty line.

ITZXITZX committed:
Merge branch 'test1'

ITZXITZX committed:
Add descriptions to methods.

Methods require documentation

ITZXITZX committed:
Save: save records into a txt file.

Data cannot be saved when users use Bob in different occasions.

Save records into a txt file. Use txt file to load records upon the
start of the program.

Records can be saved.

Let's:
- create methods to load records, save records and parse Data in Bob.
- create file format method for each task type
- get the main program to load data and save data

ITZXITZX committed:
Merge branch 'branch-Level-7'

ITZXITZX committed:
Store date as java.time.LocalDate objects.

ITZXITZX committed:
Merge branch 'branch-Level-8'

ITZXITZX committed:
Remove unused methods.

Clarenceeey commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+
+public class Deadline extends Task {
+
+    protected String by;
```
Name 'by' doesnt sound very descriptive of the variable being stored

ITZXITZX replied:
Thanks! I have changed the name.

Clarenceeey commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,58 @@
+
+    // Returns the letter representing deadline.
+    @Override
+    public String taskLetter() {
```
Wow i love how you named this method! It is very descriptive!

Clarenceeey commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,47 @@
+        super(description);
+        this.startDay = startDay;
+        this.startTime = startTime;
+        this.endTime = endTime;
```
This variable name tells me that this variable is the endTime because it is named endTime!

Clarenceeey commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,47 @@
+     * Returns a string representation of the file format in which we store the Event.
+     */
+    @Override
+    public String fileFormat () {
```
Name of method tells me exactly what it should return. Great job!

Clarenceeey reviewed (commented):
LGTM 👍

ITZXITZX committed:
Added Parser, TaskList, Storage and Ui classes.

ITZXITZX committed:
Update the methods of the new classes, remove old methods.

ITZXITZX committed:
Merge branch 'branch-A-MoreOOP'

ITZXITZX committed:
Update error messages for event.

ITZXITZX committed:
Put files into packages.

ITZXITZX committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ITZXITZX committed:
Build Gradle

ITZXITZX committed:
Add 2 JUnit test.

ITZXITZX committed:
Complete A-Jar task.

ITZXITZX committed:
Remove unnecessary imports.

ITZXITZX committed:
Add find function.

ITZXITZX committed:
Update javadoc for Bob and Ui class.

ITZXITZX committed:
Update javadoc for all task files.

ITZXITZX committed:
Complete javadoc for all files in bob.

ITZXITZX committed:
Add getTaskListItem as an overridden method in task's subclasses.

ITZXITZX committed:
Update java coding standards.

ITZXITZX committed:
Merge branch 'branch-A-JavaDoc'

ITZXITZX committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/bob/Bob.java
#	src/main/java/bob/data/tasks.txt
#	src/main/java/bob/task/Task.java
#	src/main/java/bob/task/TaskList.java
#	src/main/java/bob/task/Todo.java
#	src/main/java/bob/ui/Ui.java

OrangeCatLoves commented on src/test/java/bob/task/TaskListTest.java:
```diff
@@ -0,0 +1,39 @@
+    private ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();
+    private TaskList taskList;
+
+    @BeforeEach
+    public void setUp() {
+        System.setOut(new PrintStream(outputStreamCaptor));
+        ArrayList<Task> tasks = new ArrayList<>();
+        tasks.add(new Todo("Code"));
+        tasks.add(new Deadline("Submit quiz", LocalDate.parse("2024-09-10")));
+        tasks.add(new Event("Birthday", "Monday", "5pm", "9pm"));
+        taskList = new TaskList(tasks);
+    }
```
Hi. setUp() might not be informative enough as to what the test case is suppose to do. Perhaps you could rename it to 
initialiseTaskListTest().

ITZXITZX replied:
Oh that's a great suggestion, thanks! I have changed it.

OrangeCatLoves commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,91 @@
+    public void run() {
+        ui.showWelcome();
+        Scanner scanner = new Scanner(System.in);
+        String input = scanner.nextLine().trim(); //input with NO whitespace in front/back
+        while (!input.equals("bye")) {
+            String[] inputWords = input.split("\s+");
+            String keyword = inputWords[0];
+
+            switch (keyword) {
+            case "list":
+                taskList.listRecords();
+                break;
+            case "mark":
+                taskList.updateMark(input, inputWords, true);
+                break;
+            case "unmark":
+                taskList.updateMark(input, inputWords, false);
+                break;
+            case "delete":
+                taskList.delete(input);
+                break;
+            case "event":
+                taskList.addTask(input, inputWords);
+                break;
+            case "deadline":
+                taskList.addTask(input, inputWords);
+                break;
+            case "todo":
+                taskList.addTask(input, inputWords);
+                break;
+            case "find":
+                taskList.find(input);
+                break;
+            default:
+                Ui.requestValidCommand();
+            }
+            taskList.saveRecords(storage);
+            input = scanner.nextLine().trim();
+        }
+        Ui.showGoodBye();
```
Nice that you condense the cases for different Tasks into different cases! Another way would be to create a new Command interface and then create a new Command class that implements this command. For example, you could create deleteCommand class that implements Command.

ITZXITZX replied:
I like your suggestion, thanks!

OrangeCatLoves commented on src/main/java/bob/storage/Storage.java:
```diff
@@ -0,0 +1,124 @@
+package bob.storage;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+
+import bob.task.Deadline;
+import bob.task.Event;
+import bob.task.Task;
+import bob.task.Todo;
```
Good that you leave spaces between different imports to demarcate the separation! I realised that I did not do this - will go change later haha

OrangeCatLoves reviewed (commented):
Code is largely readable! Good job

vicelyav commented on src/main/java/bob/storage/Storage.java:
```diff
@@ -0,0 +1,124 @@
+        String[] parts = lineInput.split(" \\| ");
+        boolean isDone = (parts[1].equals("1")) ? true : false;
+
+        switch (parts[0]) {
```
Good that you followed the switch-case indentation :)

vicelyav commented on src/main/java/bob/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+    public void markTask(boolean value) {
+        if (value) {
+            this.isDone = true;
+            String finishedMarking = "OK, I've marked this task as done:\n\t"
```
This is a very clear way of writing the formatted String. Good idea!

vicelyav commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,91 @@
+import java.util.Scanner;
+
+import bob.storage.Storage;
+import bob.task.Task;
```
Good job on organising the import statements into neat blocks

vicelyav commented on src/main/java/bob/task/TaskList.java:
```diff
@@ -0,0 +1,332 @@
+            String keyword = inputWords[0];
+            Task newTask = getTask(keyword, input); //initialise the exact Task class
+            String immediateAdd = "Got it. I've added this task:\n\t"
+                    + "  ["
```
Good that you always remember when to leave white spaces especially for operators.

vicelyav commented on src/main/java/bob/task/TaskList.java:
```diff
@@ -0,0 +1,332 @@
+
+    /**
+     * Returns the String representation of the task description only.
+     * @param input input by the user.
```
I think you could leave an empty line between the Javadoc description and the parameter

ITZXITZX replied:
Oh yes, I should, thanks!

vicelyav reviewed (commented):
Overall really good job in following the coding standard. I will definitely get inspiration from your very neat code!

ITZXITZX committed:
Change names to reflect more accurate meaning.

ITZXITZX committed:
Add Command classes and subclasses.

ITZXITZX committed:
Update ListCommand and MarkCommand.

ITZXITZX committed:
Implement DeleteCommand.

ITZXITZX committed:
Update FindCommand.java

ITZXITZX committed:
Remove redundant taskList field.

ITZXITZX committed:
Update all remaining Command subclasses.

ITZXITZX committed:
Able to gradlew build.

ITZXITZX committed:
Complete GUI for Bob.

ITZXITZX committed:
Merge branch 'branch-Level-10'

ITZXITZX committed:
Fix: ensure "bye" command terminates the GUI application.

The "bye" command did not terminate the GUI as expected.

Add a check to verify the termination status after each user input. If
the user enters "bye," the GUI will close properly.

ITZXITZX committed:
Add Assertions.

ITZXITZX committed:
Update file path of gui to the correct file path.

ITZXITZX committed:
Merge pull request #2 from ITZXITZX/branch-A-Assertions

Add Assertions.

ITZXITZX committed:
Merge branch 'master' of https://github.com/ITZXITZX/ip

* 'master' of https://github.com/ITZXITZX/ip:
  Update file path of gui to the correct file path.
  Add Assertions.
  Fix: ensure "bye" command terminates the GUI application.

ITZXITZX committed:
Merge branch 'master' into branch-A-CodeQuality

ITZXITZX committed:
Remove unused methods and fields.

ITZXITZX committed:
Codes does not SLAP.

Code is difficult to read.

Refactor affected methods across all classes.

This safely neatens the code without modifying the code itself.

Let's refactor all methods that do not follow SLAP.

ITZXITZX committed:
Merge branch 'branch-A-CodeQuality'

ITZXITZX committed:
Tag Command: tag tasks

Task can not be tagged.

Tag help to mark task in a more customizable way.

Let's,
* update task to accept tags
* add a tag command
* modify saving and loading of tagged commands

ITZXITZX committed:
Untag Command: Removes the tag of the task.

Unable to remove tag from task.

Included a untag command.

ITZXITZX committed:
Merge branch 'branch-C-Tagging'

ITZXITZX committed:
Add README

ITZXITZX committed:
Update Command's execute method

ITZXITZX committed:
Set title for gui stage

ITZXITZX committed:
Change colors of gui background, text, button and scroll bar

ITZXITZX committed:
Add welcome message upon start of gui

ITZXITZX committed:
Add Ui.png

ITZXITZX committed:
Fix remaining bugs

ITZXITZX committed:
Remove unused methods, classes

ITZXITZX committed:
Complete A-UserGuide

ITZXITZX committed:
Include acknowledgements in README

ITZXITZX committed:
Add ParserTest test cases

ITZXITZX committed:
Add TaskListTest cases

ITZXITZX committed:
Update file path

ITZXITZX committed:
Complete A-Release
